### PR TITLE
Override an unversioned object originOp when moving it to a version

### PR DESCRIPTION
**Pull request by delthas**
This prevents triggering a possible s3:ObjectCreated:Put bucket notification when moving this object to a version, when deleting it (when creating a deletion marker).

Issue: CLDSRV-816

### CODE REVIEW

**Comment by delthas** [comment]
Tested locally and with the full workflow on an ARTESCA, working for both manual deletion and expiration workflows.

File: lib/api/apiUtils/object/versioning.js [hunk 1]
```diff
@@ -98,6 +98,7 @@ function _storeNullVersionMD(bucketName, objKey, nullVersionId, objMD, log, cb)
             isNull2: true,
         });
     }
+    nullVersionMD.originOp = 's3:ObjectNullVersionStoredAsNewVersion';
```
Comment (francoisferrand) [inline lib/api/apiUtils/object/versioning.js]: maybe something shorter: `StoreNullVersion` ?
(minimal optimization, but anything we add ends up pilling in metadata.... and with billions of objects, it quickly adds up)

File: lib/api/apiUtils/object/versioning.js
Comment (francoisferrand) [inline lib/api/apiUtils/object/versioning.js:83]: in which situation is this function called? i.e. 
- if it called only when we have a "master-only" version, to "migrate it" to a proper version → fine
- if however it may be called in other cases, like when (over)writing an object (for exemple in version-suspended buckets ?), then we must keep the existing originOp in that case

Did you perform such kind of "impact analysis"? Are we sure this is only called to perform an _extra_ put, for migration purpose, and never to create/replace the new version itself?

→ would be great to check the tests, and (if needed) update them to check that the proper originOp is set in each case (`putObject`, `copyObject`, `putMultipart` and `backeatPutMetadata` in the different combination of versioned mode & pre-existing object, ...)

File: lib/api/apiUtils/object/versioning.js [hunk 2]
```diff
@@ -98,6 +98,7 @@ function _storeNullVersionMD(bucketName, objKey, nullVersionId, objMD, log, cb)
             isNull2: true,
         });
     }
```
Comment (francoisferrand) [inline lib/api/apiUtils/object/versioning.js:100]: this was missed in the existing tests : so we should introduce -or update- a test (unit and/or functional, in line with how it is already tested) to make sure this behavior will stay.

Comment (delthas) [reply lib/api/apiUtils/object/versioning.js:100]: OK, added a test that checks for versionId, isNull, and originOp. https://github.com/scality/cloudserver/pull/6036/files#diff-dffdecd8dd3ea359b81fdebca3575a93b16e025dcfca47e1f609282f31b70764R425

Do we want to test anything else in that test? (Apart from check the existing originOp in existing put/copy/...)

File: lib/api/apiUtils/object/versioning.js
Comment (delthas) [reply lib/api/apiUtils/object/versioning.js:83]: OK, added some more tests.

File: tests/unit/api/objectPut.js [hunk 3]
```diff
@@ -980,6 +980,30 @@ describe('objectPut API with versioning', () => {
         });
     });
 
+    it('should set originOp when moving null version', done => {
```
Comment (SylvainSenechal) [inline tests/unit/api/objectPut.js]: I think it's worth it to document more either this test or the one in objectDelete.js. 
Here you added a comment "M was moved to V", I understand it because I read the Jira ticket where you give a lot of context, but someone else probably wouldn't understand. 

This new StoreNullVersion doesn't look very documented either, so this test could be a an opportunity to write a few line about the bug that happened before and why we added StoreNullVersion to fix it

**Review by SylvainSenechal — APPROVED** [review]

File: tests/unit/api/objectDelete.js [hunk 4]
```diff
@@ -371,3 +374,59 @@ describe('objectDelete API', () => {
         });
     });
 });
+
+
```
Comment (francoisferrand) [inline tests/unit/api/objectDelete.js]: nit: remove extra empty line

```suggestion
```

File: tests/unit/api/objectDelete.js [hunk 5]
```diff
@@ -371,3 +374,59 @@ describe('objectDelete API', () => {
...
+    });
+    const testDeleteRequest = new DummyRequest({
+        bucketName,
+        namespace,
+        objectKey,
+        headers: {},
+        url: `/${bucketName}/${objectKey}`,
+    });
+
+    it('should store modified originOp when moving object null version', done => {
```
Comment (francoisferrand) [inline tests/unit/api/objectDelete.js]: maybe better describe the scenario, to explain this happens when putting a delete marker on top of a "master-only null version"


```suggestion
    it('should store modified originOp when storing null object due to object marker', done => {
```

...or we could make this more general, and consider this test actually tests this overall process of "upgrading" the null-version object (in case an object marker is put on top of it) : i.e. we do not just want to verify that the object has the proper originOp, but also that it has all proper metadata and "null" version id (which is mostly what you do here, actually)

File: tests/unit/api/objectPut.js (hunk 3 above)
Comment (francoisferrand) [reply tests/unit/api/objectPut.js]: > This new StoreNullVersion doesn't look very documented either, so this test could be a an opportunity to write a few line about the bug that happened before and why we added StoreNullVersion to fix it

this is not a bug, this is a "known" corner-case when moving from the "master-only bucket format" (non-versioned) to the "versioned bucket format". And this is documented, in `docs/ARCHITECTURE.rst`, in section "Preserving Existing Null Versions in Versioning-Enabled Bucket".

File: tests/unit/api/objectDelete.js [hunk 6]
```diff
@@ -371,3 +374,59 @@ describe('objectDelete API', () => {
...
+            next => objectDelete(authInfo, testDeleteRequest, log, next),
+            async () => {
+                const calls = metadataswitch.putObjectMD.getCalls();
+                sinon.assert.calledWith(calls[calls.length - 2],
+                    bucketName, objectKey, sinon.match({
+                        versionId: sinon.match.truthy,
+                        isNull: true,
+                        originOp: 's3:StoreNullVersion',
+                    }), any, any, any);
+            },
```
Comment (francoisferrand) [inline tests/unit/api/objectDelete.js:428]: just like the other test, I guess we should also check that the second call created a delete marker with the expected originOp (`s3:ObjectRemoved:DeleteMarkerCreated`)

File: tests/unit/api/objectPut.js [hunk 7]
```diff
@@ -980,6 +980,30 @@ describe('objectPut API with versioning', () => {
...
+            async () => {
+                // M was moved to V, with originOp overridden to prevent bucket notifications.
+                const calls = metadata.putObjectMD.getCalls();
+                sinon.assert.calledWith(calls[calls.length - 2],
+                    bucketName, objectName, sinon.match({
+                        originOp: 's3:StoreNullVersion',
+                    }), any, any, any);
+            },
+            async () => {
+                // New V was created with the right originOp.
```
Comment (francoisferrand) [inline tests/unit/api/objectPut.js]:
```suggestion
                // New version document was created with the right originOp.
```

File: tests/unit/api/objectPut.js [hunk 8]
```diff
@@ -980,6 +980,30 @@ describe('objectPut API with versioning', () => {
...
     });
 
+    it('should set originOp when moving null version', done => {
+        async.series([
+            next => bucketPut(authInfo, testPutBucketRequest, log, next),
+            next => objectPut(authInfo, testPutObjectRequest, undefined, log, next),
+            next => bucketPutVersioning(authInfo, enableVersioningRequest, log, next),
+            next => objectPut(authInfo, testPutObjectRequest, undefined, log, next),
+            async () => {
+                // M was moved to V, with originOp overridden to prevent bucket notifications.
```
Comment (francoisferrand) [inline tests/unit/api/objectPut.js]:
```suggestion
                // Old master-only document was moved to a proper version, with originOp overridden to prevent unexpected bucket notifications.
```

**Review by francoisferrand — APPROVED** [review]

File: tests/unit/api/objectPut.js (hunk 3 above)
Comment (delthas) [reply tests/unit/api/objectPut.js]: Slightlly clarified the test name.

File: tests/unit/api/objectDelete.js (hunk 5 above)
Comment (delthas) [reply tests/unit/api/objectDelete.js]: OK, clarified the test name accordingly.
